### ISSUE TITLE
2.0: Android app crashed while running with 1 million docs

### ISSUE THREAD

**Issue by sridevi-15** [couchbase/couchbase-lite-android#1555]
CBL version : CBL 2.0 with DB22
Platform : Android

Steps to reproduce : 
1. Set up sync-gateway
2. Create docs in batch in CBL
3. Start replication with continuous true and with push_pull replication 

Result  : 
App crashes

Expected : 
App should continue to replicate successfully

Logs in gist : https://gist.github.com/sridevi-15/4a96bdf5df589824b174686e16fa2c35

Code for this test  :  https://github.com/couchbaselabs/mobile-testkit/blob/feature/cbl20-ver1/apps/testkit.java/testkit.java.android/Tests/AndroidClient/app/src/main/java/com/couchbase/androidclient/MainActivity.java

**Comment by sridevi-15** [comment]
@hrajput89 , I created this ticket as we have reproducible steps in Android. Not sure the exact steps and where exactly it failed for you. I assumed the steps I have written is right. Please add if anything is missing

**Comment by hideki** [comment]
@sridevi-15 Do you have the crash log from logcat?

**Comment by hideki** [comment]
Note to self:

Is this root cause?
```
A/art: art/runtime/runtime.cc:427] JNI ERROR (app bug): weak global reference table overflow (max=51200)
```

**Comment by hideki** [comment]
Hi @sridevi-15 and @hrajput89 
Could you please provide Android device info and Android API version?
Thanks!

**Comment by hideki** [comment]
@hrajput89 and @sridevi-15 

Inserting 1M documents in one batch is impossible!

https://github.com/couchbaselabs/mobile-testkit/blob/feature/cbl20-ver1/apps/testkit.java/testkit.java.android/Tests/AndroidClient/app/src/main/java/com/couchbase/androidclient/MainActivity.java#L94-L105

**Comment by hideki** [comment]
@hrajput89 and @sridevi-15

What version of SG build are you using?

**Comment by hideki** [comment]
Hi @hrajput89,

Can you provide full test env information?

Thanks

**Comment by ssmotra** [comment]
@hrajput89 - what's the size of the DB?

**Comment by sridevi-15** [comment]
SG build is 759. running on Android emulator. will make a change to save document 1K for each batch and run it .

**Comment by hideki** [comment]
NOTE:
QA uses Gennymotion for testing.
Gennymotion setting is `google nexus 9 with 7.0.0 API for emulator is 24`.

**Comment by hideki** [comment]
@sridevi-15, @hrajput89 

We have some questions about this test case.

- Is this the regression from the previous build?
- Is this the regression from the v1.4.1? (or Is this new test case for 2.0?)
- Could you please describe detailed test environment?
  - Sync Gateway version
  - Emulator information: Google Emulator or GennyMotion
  - Android version (Android API version)
  - VM environment

**Comment by sridevi-15** [comment]
@hideki : 
1. This is first time ran with 1Million docs. This test was ran in previous build with small set and did not run with 1Million in previous builds.
2. This is not regression from 1.4.1.There was no test with this scenarion on 1.4.1
3. sg version : 759
     emulator : geny motion with API 24 . Google Nexus 9
     Andriod API : 24 
    VM : Centos which sync-gateway set up 
4. Num of docs : Tested with 100k docs , it worked fine, 1 Million docs failed. 
   scenarioRunTimeMinutes : This is the number of minutes it will run to update docs randomly . This should not matter with crate 1 Million docs. 

We tested this scenario with DB22 first build. Can try with latest build of DB22 by making 100k chunks while save documents in batch?
Note : Creating 100k docs in batch already worked. Making chunks into 100k for 1Million set should work . Will update this ticket once I modify the test and if it works

**Comment by hideki** [comment]
Note to Self: Analysis of this issue

### 1. What potentially caused the crash from the log. 
From the log, `weak global reference table overflow` caused the crash. From following info, the native codes reference to many `java.lang.Thread` instances.
```
A/art: art/runtime/runtime.cc:427] JNI ERROR (app bug): weak global reference table overflow (max=51200)
A/art: art/runtime/runtime.cc:427] weak global reference table dump:
A/art: art/runtime/runtime.cc:427]   Last 10 entries (of 51200):
A/art: art/runtime/runtime.cc:427]     51199: 0x13eeb310 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51198: 0x13eeb160 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51197: 0x13d40e50 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51196: 0x13d40d30 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51195: 0x13d40a60 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51194: 0x13d408b0 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51193: 0x13d40700 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51192: 0x13d40550 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51191: 0x13d403a0 java.lang.Thread
A/art: art/runtime/runtime.cc:427]     51190: 0x13d400d0 java.lang.Thread
A/art: art/runtime/runtime.cc:427]   Summary:
A/art: art/runtime/runtime.cc:427]         1 of android.support.v4.util.SparseArrayCompat
A/art: art/runtime/runtime.cc:427]         1 of android.arch.lifecycle.LifecycleRegistry
A/art: art/runtime/runtime.cc:427]         1 of com.couchbase.lite.DatabaseConfiguration
A/art: art/runtime/runtime.cc:427]         1 of android.support.v4.util.SimpleArrayMap
A/art: art/runtime/runtime.cc:427]         1 of com.couchbase.lite.Replicator
A/art: art/runtime/runtime.cc:427]         1 of android.support.v4.app.FragmentActivity$1
A/art: art/runtime/runtime.cc:427]         1 of com.couchbase.androidclient.MainActivity
A/art: art/runtime/runtime.cc:427]         1 of android.support.v4.app.FragmentController
A/art: art/runtime/runtime.cc:427]         1 of com.couchbase.lite.Database
A/art: art/runtime/runtime.cc:427]         1 of android.support.v7.app.AppCompatDelegateImplN
A/art: art/runtime/runtime.cc:427]         1 of okio.AsyncTimeout$Watchdog
A/art: art/runtime/runtime.cc:427]       680 of java.lang.Thread (680 unique instances)
```
https://gist.github.com/sridevi-15/4a96bdf5df589824b174686e16fa2c35#file-gistfile1-txt-L311-L336


### 2. What creates the reference to `java.lang.Thread` from JNI, native codes?

`AttachCurrentThread()` and `DetachCurrentThread()` creates to the `java.lang.Thread` and releae it.

Read *Threads* section in JNI Tips document.
https://developer.android.com/training/articles/perf-jni.html#threads

The example in CBL Jni codes. To call Java codes from JNI (native codes), JNI code need to attach current thread. And after callback function finished, it detaches from the thread. Between attach and detach, it is bined to one instance of `java.lang.Thread`. If thow methods call this concurrently, it attach two Thread instances. 
```c++
        if (gJVM->AttachCurrentThread(&env, NULL) == 0) {
            env->CallStaticVoidMethod(cls_C4Replicator, m_C4Replicator_statusChangedCallback,
                                      (jlong) repl,
                                      toJavaObject(env, status));
            if (gJVM->DetachCurrentThread() != 0)
```
### 3. Why are many Thread instances referenced from JNI evn though JNI (native codes) detachs it immediatly.

Callback is processed with UI/Main thread.
```java
this.handler = new Handler(Looper.getMainLooper());
...
...
            public void statusChanged(final C4Replicator repl, final C4ReplicatorStatus status, final Object context) {
                Log.i(TAG, "C4ReplicatorListener.statusChanged() status -> " + status);
                final Replicator replicator = (Replicator) context;
                if (repl == replicator.c4repl) {
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            replicator.c4StatusChanged(status);
                        }
                    });
                }
            }
```
https://github.com/couchbase/couchbase-lite-android/blob/feature/2.0/shared/src/main/java/com/couchbase/lite/Replicator.java#L224
https://github.com/couchbase/couchbase-lite-android/blob/feature/2.0/shared/src/main/java/com/couchbase/lite/Replicator.java#L395-L406

And also test code blocks Main/UI thread. `Activity.onStart()` is executed from main/UI thread.
```java
    @Override
    protected void onStart() {
       #### All test codes are implemented in this method #####
    }
```
https://github.com/couchbaselabs/mobile-testkit/blob/feature/cbl20-ver1/apps/testkit.java/testkit.java.android/Tests/AndroidClient/app/src/main/java/com/couchbase/androidclient/MainActivity.java#L85-L154

As result, all callbacks from replicator blocked because onStart() method blocks Main/UI thread.

### 4. How to solve this issue
This issue is alredy know, and scheduled to be fixed by DB023.
2.0: Replicator uses Main/UI thread for notification. Should use background thread 
https://github.com/couchbase/couchbase-lite-android/issues/1560

**Comment by hideki** [comment]
Note to self:
With modifying callback method thread, equivalent test codes can handle 1M document push and pull replication.
https://github.com/hideki/Issue1555/blob/master/app/src/main/java/com/couchbase/issue1555/MainActivity.java

**Comment by hideki** [comment]
`Database.inBatch()` method calls callback from same thread that inBatch() is executed.
But test code keeps block UI thread by onStart().
This is another point that causes the weak global reference table over flow.

```java
    public void inBatch(Runnable runnable) throws CouchbaseLiteException {
        ...
        ...
        postDatabaseChanged();
    }
```
https://github.com/couchbase/couchbase-lite-android/blob/feature/2.0/shared/src/main/java/com/couchbase/lite/Database.java#L351

```java
        database.addChangeListener(new DatabaseChangeListener() {
            @Override
            public void changed(DatabaseChange change) {
                Log.i("Database change listener", "%s", change);
            }
        });
```
https://github.com/couchbaselabs/mobile-testkit/blob/feature/cbl20-ver1/apps/testkit.java/testkit.java.android/Tests/AndroidClient/app/src/main/java/com/couchbase/androidclient/MainActivity.java#L58-L63

**Comment by hrajput89** [comment]
I tested the app again with 1M docs with batch of 1000, iterating for 1000 docs at a time, as suggested by Hideki. 

Test Environment - Android

Genymotion emulator google Pixel XL7.1.0 - API 25

SG -
  - OS : CENTOS7
  - mode : cc
  - build : 2.0.0-749

CBS -
  - OS : Centos7
  - build : 5.0.0-3519

Steps to Reproduce - 
1. Set cluster with SG and CBS in cc mode.
2. open AndroidClient app in android studio and modify below variables -
 - final static int BATCH_SIZE = 1000;
 - final static int ITERATION = 1000;
 - numOfDocs = 1000000;
 - scenarioRunTimeMinutes = 1;
 - syncGatewayURL = "ws://172.16.1.107:4985/db/";
3. Run the app in debug mode on Genymotion simulator


Below are the logs attached for the run -
[sgcollect_info.zip](https://github.com/couchbase/couchbase-lite-android/files/1676879/sgcollect_info.zip)
[system_test_logs_for_297.log](https://github.com/couchbase/couchbase-lite-android/files/1676880/system_test_logs_for_297.log)


Here is the app code available - https://github.com/couchbaselabs/mobile-testkit/tree/feature/cbl20-ver1/apps/testkit.java/testkit.java.android

**Comment by hideki** [comment]
@hrajput89 

I am wondering if debug mode on Genymotion simulator could cause the problem. 

Please check following stack overflow comment.

Can you test this test case with non-debug mode?

https://stackoverflow.com/questions/40028626/jni-error-app-bug-weak-global-reference-table-overflow-why

**Comment by hideki** [comment]
Information  from @hrajput89,

```
app didn't crash for 1M doc when ran on android emulator without debug mode
```
Link to log
https://gist.github.com/hideki/381fd456e59bed3e6c50d208390caf33

**Comment by djpongh** [comment]
@hideki what happens with your latest build when this test is run in debug mode? Just trying to understand what your fix does.

**Comment by hrajput89** [comment]
I ran the System testing app for 1M doc (1000 doc in batch iterating for 1000 times) for an hours without the debug mode on Genymotion emulator. App ran successfully. Below are the logs attached


[System_Testing_app.log](https://github.com/couchbase/couchbase-lite-android/files/1679145/System_Testing_app.log)

**Comment by hideki** [comment]
Hi @djpongh 

All actions we take based on [a log file](https://gist.github.com/sridevi-15/4a96bdf5df589824b174686e16fa2c35).
- With 2.0.0-297, I fixed the problem that I described in my [comment](https://github.com/couchbase/couchbase-lite-android/issues/1555#issuecomment-361499745).
- I asked @hrajput89 to [reduce batch size to 1K from 1M](https://github.com/couchbase/couchbase-lite-android/issues/1555#issuecomment-361379929), and iterate 1K times to same 1M docs.
- Also, I asked @hrajput89 to not attach debug to Emulator. Debugger causes weak global reference leak based on [StackOverFlow](https://stackoverflow.com/questions/40028626/jni-error-app-bug-weak-global-reference-table-overflow-why).

We sequentially applied the changes. From this, I believe debugger modes causes the weak global reference leak and crash by it. Other two changes are an important change, but the primary one is debugging mode.
